source/manuals/programming-languages/python/python.html.md.erb

```diff
@@ -1,106 +1,150 @@
-    message][dm-deps-commit].
+These recommendations apply wherever you need a reproducible set of dependencies, such
+as a complete web application, perhaps with many dependencies and sub-dependencies. It also
+applies to a collection of scripts that are deployed into the cloud and run automatically
+(for example, batch jobs).
 
-    **Recommendation**: Put your top-level requirements in a
-    `requirements-app.txt` file and use [this Makefile
-    script][dm-deps-commit-makefile] to generate a fully specified
-    `requirements.txt`, which is used whenever you need to pip-install the
-    dependencies. The script creates a temporary virtualenv, pip-installs the
-    `requirements-app.txt` and pip-freezes the result as `requirements.txt`.
+A good strategy for specifying your application's Python dependencies has two desirable
+characteristics - they should be:
 
-    **Recommendation**: Use `requirements-dev.txt` for any dependencies used in
-    development/test but not production environments.
+1. Reproducible (predictable)
 
+    Pin your application's full dependencies – specific versions, rather than ranges – or you'll
+    get unpredictability between your dev environment and other environments. You want a
+    new starter to avoid small hard-to-spot problems. And you want parity between what you
+    test locally, what is tested by CI, and what you deploy, or you risk new issues appearing on
+    a live server. Additionally these things can be hard to diagnose.
 
 2. Kept up-to-date
 
-    Security issues are found in libraries, so it is important to choose
-    libraries that are maintained and to ensure your team has a strategy to
-    ensure security updates are installed without significant delay.
+    Security issues are found in libraries, so it is important to choose libraries that are
+    maintained and to ensure your team has a strategy to ensure security updates are installed
+    without significant delay. The [how to manage third party software dependencies][gds-way-deps] section
+    gives further context and discusses tools that can help, such as [snyk.io][].
+
+Your README should document an easy-to-follow process by which all your Python
+dependencies can be upgraded to get bug fixes and security fixes, without introducing
+breaking changes to your build.
+
+Your pinned dependencies should be fully specified in a file called `requirements.txt`, and checked
+into your version control system (VCS). For projects with only a small number of dependencies, maintaining
+this manually (for example, installing with `pip install`, then using `pip freeze`) may be adequate.
+
+For larger projects, GDS recommends the following approach, which was developed after various issues
+were found in current mainstream tooling (see [this commit message][dm-deps-commit], and note also the
+[lack of support for specifying VCS dependencies](https://github.com/jazzband/pip-tools/issues/355)
```

Comment:
Given that this is a moment time in evaluation of a set of tools, some of which are quite new, I would prefer for this section to explicitly mention specific tools that were evaluated (eg pipenv version such and such) and summarise the reasons against adoption at this time (eg lack of support for VCS dependencies), instead of linking out to a decision made for a particular project.

I think it's important that people who are thinking of deviating from the recommendation can make an informed decision, based on what we know now and any additional information available to them at the time.

Similarly, people reviewing this document in X months time should be able to quickly evaluate whether the conclusion still makes sense given how the tools have changed since it was written.

Reply:
I see your point, however I don't know how much detail is supposed to go into these manuals; the small amount I know about previous efforts in this area is that there have been a lot of experiments and a lot of dead ends, so documenting everything could take a lot of words!

Reply:
A suggestion from @benvand, maybe the best way to deal with this issue is to put some form of words at the top of the page saying `If you have an issue with or want to deviate from the recommendations in this manual, please contact #python-community on Slack in the first instance.`

Reply:
So that last point is one reason I added that link to a specific github issue. But in truth we haven't fully re-reviewed the cluster**** that is the official Python packaging toolset: it's not a moment in time evaluation, it's a collection of a couple of evaluations over a period of a year.

This time, I've done some evaluation of `pip-compile`, and that issue/feature request seems me to be the biggest blocker. We should take another look at `pipenv`, but I'd wager @allait's previous comments do still apply (see the referenced PR).

I personally suspect that `pipenv` simply tries to do too much (especially given that containerisation options move along in their own way), and that what we really want is a working `pip-compile`.

@benvand has already started a [reimplementation in Python](https://github.com/alphagov/digitalmarketplace-utils/pull/473/files) of our current approach for use in Digital Marketplace. Once that's actually been rolled out anywhere then we could certainly reference it from here, and I have a few ideas about how to improve it (slightly).

All this feels interim, and I'd definitely encourage anyone that has the energy to consider either fully re-evaluating `pipenv` on a complex project like Digital Marketplace or perhaps Notify, or look at hacking on that feature in `pip-compile`.

I'm not sure there's a way we can succinctly document current practice in such a way as to much assist with that 'quickly evaluate' thing, but if you can reword at all in that direction then go for it 👍